### PR TITLE
Move monitor from per-entry to cache

### PR DESCRIPTION
**Pull request by rolandshoemaker**
Is this along the lines of what you were thinking @jsha?

### CODE REVIEW

File: cache.go [hunk 1]
```diff
@@ -34,8 +34,10 @@ type cache struct {
 	mu        *sync.RWMutex
 }
 
-func newCache(log *Logger) *cache {
-	return &cache{log, make(map[string]*Entry), make(map[[32]byte]*Entry), new(sync.RWMutex)}
+func newCache(log *Logger, monitorTick time.Duration) *cache {
+	c := &cache{log, make(map[string]*Entry), make(map[[32]byte]*Entry), new(sync.RWMutex)}
```
Comment (jsha) [inline cache.go]: line per entry, with names, wouldbe better here.

File: cache.go [hunk 2]
```diff
@@ -139,11 +140,32 @@ func (c *cache) remove(name string) error {
 	return nil
 }
 
+func (c *cache) monitor(tick time.Duration) {
+	ticker := time.NewTicker(tick)
+	for {
+		select {
```
Comment (jsha) [inline cache.go]: No need for a select if you're only selecting on one thing.

File: cache.go [hunk 3]
```diff
@@ -139,11 +140,32 @@ func (c *cache) remove(name string) error {
...
+		case <-ticker.C:
+			c.mu.RLock()
+			defer c.mu.RUnlock()
+			for _, entry := range c.entries {
+				if entry.timeToUpdate() {
+					go func() {
+						err := entry.refreshResponse()
+						if err != nil {
+							entry.log.Err("Failed to refresh response: %s", err)
```
Comment (jsha) [inline cache.go]: Why not do this inside refreshResponse (or a wrapper function)? Also, then you can use `e.err` (which you can do here too, but seems more natural there).

**Comment by jsha** [comment]
Yep, this is what I was thinking of!

**Comment by rolandshoemaker** [comment]
I think I covered everything, any other thoughts before I merge?

File: cache.go [hunk 4]
```diff
@@ -139,11 +144,22 @@ func (c *cache) remove(name string) error {
 	return nil
 }
 
+func (c *cache) monitor(tick time.Duration) {
+	ticker := time.NewTicker(tick)
+	for range ticker.C {
+		c.mu.RLock()
```
Comment (jsha) [inline cache.go:150]: This is a long time to hold this lock, even if it is a read lock. Maybe hold the lock, make a copy of entries, and then unlock while you do the fetches?

Comment (rolandshoemaker) [inline cache.go:150]: A goroutine is spawned for each `refreshAndLog` so the time the lock is held should be quite small. The lock is only really needed to make sure that no entries are removed while iterating over `c.entries` so its okay to release it before all the refreshes actually finish.

Comment (jsha) [inline cache.go:150]: Right, I forgot that.

**Comment by jsha** [comment]
LGTM

**Pull request merged**